File: src/content_factory/models/campaign.py
"""Multi-format campaign fan-out: shorts, prompt packs, ratios."""

from __future__ import annotations

import enum
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

from .common import utcnow


class HybridAssetType(enum.StrEnum):
    """Visual asset classification for hybrid documentary production."""

    AI_RECONSTRUCTION = "ai_reconstruction"  # Kling / Veo / Wan 2.1
    HISTORICAL_PHOTO = "historical_photo"  # Archival photo / Public domain
    DYNAMIC_MAP = "dynamic_map"  # Route map / Geo animation
    DECLASSIFIED_DOC = "declassified_doc"  # Newspapers / Archives / Dossiers
    TECHNICAL_DIAGRAM = "technical_diagram"  # Cross-section / Schematics
    KINETIC_MOTION = "kinetic_motion"  # Typography / Infographics


class HybridAssetRatio(BaseModel):
    """Target percentage allocation to ensure historical credibility."""

    ai_reconstruction: int = 30
    historical_photo: int = 20
    dynamic_map: int = 15
    declassified_doc: int = 15
    technical_diagram: int = 10
    kinetic_motion: int = 10


class ShortsVariant(BaseModel):
    """A standalone vertical short video generated from the master topic."""

    id: str
    title: str
    angle: str
    hook: str
    target_duration_seconds: int = 45
    script: str
    hook_type: str = "curiosity_gap"
    scenes: list[dict[str, Any]] = Field(default_factory=list)
    video_prompts: list[str] = Field(default_factory=list)
    image_prompts: list[str] = Field(default_factory=list)
    voiceover_tone: str = "dramatic_suspense"
    call_to_action: str = ""


class PromptPack(BaseModel):
    """Optimized prompts for 3rd-party generative engines."""

    kling_veo_prompts: list[dict[str, str]] = Field(default_factory=list)
    midjourney_prompts: list[dict[str, str]] = Field(default_factory=list)
    suno_prompts: list[dict[str, str]] = Field(default_factory=list)
    elevenlabs_settings: dict[str, Any] = Field(default_factory=dict)


class MultiFormatCampaign(BaseModel):
    """Master production campaign: 1 Topic -> 1 YouTube Long + 5-10 Shorts."""

    id: str
    project_id: str
    master_topic: str
    youtube_script: str
    youtube_duration_target_seconds: int = 600  # 8-12 minutes default
    youtube_story_structure: dict[str, str] = Field(default_factory=dict)
    youtube_scenes: list[dict[str, Any]] = Field(default_factory=list)
    youtube_titles: list[str] = Field(default_factory=list)  # 5x viral titles
    youtube_description: str = ""
    youtube_chapters: list[dict[str, str]] = Field(default_factory=list)
    tiktok_hooks: list[str] = Field(default_factory=list)  # 10x retention hooks
    thumbnail_prompts: list[str] = Field(default_factory=list)
    shorts: list[ShortsVariant] = Field(default_factory=list)
    asset_ratio: HybridAssetRatio = Field(default_factory=HybridAssetRatio)
    prompt_pack: PromptPack = Field(default_factory=PromptPack)
    fact_check_summary: str = ""
    created_at: datetime = Field(default_factory=utcnow)
    updated_at: datetime = Field(default_factory=utcnow)


class CampaignGenerateRequest(BaseModel):
    """Options for generating a multi-format campaign."""

    shorts_count: int = Field(default=5, ge=1, le=10)
    youtube_target_minutes: int = Field(default=10, ge=5, le=20)
    include_prompts: bool = True


class ShortsUpdateRequest(BaseModel):
    """Payload for editing a single short in the campaign."""

    title: str | None = None
    script: str | None = None
    hook: str | None = None
    call_to_action: str | None = None

File: frontend/src/types/campaign.ts
/**
 * Multi-format campaign contract — mirrors ``models/campaign.py``.
 *
 * One root topic becomes one long-form master plus a cluster of standalone
 * vertical shorts, each with its own hook and script.
 */

/** How the media for a campaign is allocated across asset kinds. */
export interface HybridAssetRatio {
  ai_reconstruction: number;
  historical_photo: number;
  dynamic_map: number;
  declassified_doc: number;
  technical_diagram: number;
  kinetic_motion: number;
}

/** One standalone vertical short. */
export interface ShortsVariant {
  id: string;
  title: string;
  angle: string;
  hook: string;
  target_duration_seconds: number;
  script: string;
  hook_type: string;
  scenes: Array<Record<string, unknown>>;
  video_prompts: string[];
  image_prompts: string[];
  voiceover_tone: string;
  call_to_action: string;
}

/** Ready-to-paste prompt packs for external generative tools. */
export interface PromptPack {
  kling_veo_prompts: Array<Record<string, string>>;
  midjourney_prompts: Array<Record<string, string>>;
  suno_prompts: Array<Record<string, string>>;
  elevenlabs_settings: Record<string, unknown>;
}

/** A complete multi-format campaign (`MultiFormatCampaign`). */
export interface MultiFormatCampaign {
  id: string;
  project_id: string;
  master_topic: string;
  youtube_script: string;
  youtube_duration_target_seconds: number;
  youtube_story_structure: Record<string, string>;
  youtube_scenes: Array<Record<string, unknown>>;
  youtube_titles: string[];
  youtube_description: string;
  youtube_chapters: Array<Record<string, string>>;
  tiktok_hooks: string[];
  thumbnail_prompts: string[];
  shorts: ShortsVariant[];
  asset_ratio: HybridAssetRatio;
  prompt_pack: PromptPack;
  fact_check_summary: string;
  created_at: string;
  updated_at: string;
}

/** Payload for ``POST /projects/{id}/campaign/generate``. */
export interface CampaignGenerateRequest {
  shorts_count?: number;
  youtube_target_minutes?: number;
  include_prompts?: boolean;
}

/** Payload for ``PUT /projects/{id}/campaign/shorts/{short_id}``. */
export interface ShortsUpdateRequest {
  title?: string | null;
  script?: string | null;
  hook?: string | null;
  call_to_action?: string | null;
}
